src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3914,8 +3914,6 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
 
             LearnFromEqualsMethod(method, node, receiverType, results);
 
-            LearnFromCompareExchangeMethod(method, node, results);
```

Comment:
:memo: I moved/refactored this logic into the processing of assignments of parameter values to arguments. That way, `CompareExchange` behaves well when the `ref` is one of the parameters of the containing method.